### PR TITLE
[lldb][DWARFASTParserClang] Prevent unnamed bitfield creation in the presence of overlapping fields

### PR DESCRIPTION
**Pull request by Michael137**
This bug surfaced after https://github.com/llvm/llvm-project/pull/105865 (currently reverted, but blocked on this to be relanded).

Because Clang doesn't emit `DW_TAG_member`s for unnamed bitfields, LLDB has to make an educated guess about whether they existed in the source. It does so by checking whether there is a gap between where the last field ended and the currently parsed field starts. In the example test case, the empty field `padding` was folded into the storage of `data`. Because the `bit_offset` of `padding` is `0x0` and its `DW_AT_byte_size` is `0x1`, LLDB thinks the field ends at `0x1` (not quite because we first round the size to a word size, but this is an implementation detail), erroneously deducing that there's a gap between `flag` and `padding`.

This patch adds the notion of "effective field end", which accounts for fields that share storage. It is set to the end of the storage that the two fields occupy. Then we use this to check for gaps in the unnamed bitfield creation logic.

### CODE REVIEW

**Comment by Michael137** [comment]
This logic is getting very finicky, so I wouldn't be surprised if there are cases not covered by this. Will have a think

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.h [hunk 1]
```diff
@@ -258,9 +258,27 @@ class DWARFASTParserClang : public lldb_private::plugin::dwarf::DWARFASTParser {
 
 private:
   struct FieldInfo {
+    ///< Size in bits that this field occupies. Can but
```
Comment (adrian-prantl) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.h:261]:
```
/// comment
decl d;
```
or
```
decl d; ///<comment 
```

**Review by adrian-prantl — APPROVED** [review]
Once we're in heuristic territory I think we can only get a hold of this problem with lots of tests.

**Comment by Michael137** [comment]
> Once we're in heuristic territory I think we can only get a hold of this problem with lots of tests.

Added more test in the latest iteration (the ones marked FIXME were still misbehaving on top-of-tree, although in other ways...but we still asserted in the expression evaluator, so don't think we really regress with this patch).

There are some cases where it's pretty much impossible to differentiate between tail padding and unnamed bitfields. So as long as we don't emit those bitfields into DWARF I think we have to resort to this.

**Comment by adrian-prantl** [comment]
Why don't we just make the unnamed fields explicit in DWARF and solve the problem the correct way?
(We'll still needs this hack to support other/older compilers though)

**Comment by pranavk** [comment]
Hmm, this seems to be breaking LLVM build:
```
error: initialization of non-aggregate type 'FieldInfo' with a designated initializer list
 |     FieldInfo this_field_info{.is_bitfield = false};
 ```

**Comment by kazutakahirata** [comment]
> Hmm, this seems to be breaking LLVM build:
> 
> ```
> error: initialization of non-aggregate type 'FieldInfo' with a designated initializer list
>  |     FieldInfo this_field_info{.is_bitfield = false};
> ```

FWIW, with clang-16.0.6 as the host compiler, I get:

```
lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:2935:31: error: designated initializers are a C++20 extension [-Werror,-Wc++20-designator]
    FieldInfo this_field_info{.is_bitfield = false};
                              ^
1 error generated.
```

**Comment by kazutakahirata** [comment]
I've fixed the warning with 5d17293caaf0f62ea94fecc137b9b6f07c659dac.

**Comment by Michael137** [comment]
> I've fixed the warning with 5d17293caaf0f62ea94fecc137b9b6f07c659dac.
> 
> 

Thanks!

**Pull request merged**